File: bugsnag-android-performance-impl/src/testFixtures/java/com/bugsnag/android/performance/test/TestSpanFactory.java
package com.bugsnag.android.performance.test;

import com.bugsnag.android.performance.SpanKind;
import com.bugsnag.android.performance.internal.SpanCategory;
import com.bugsnag.android.performance.internal.SpanImpl;
import com.bugsnag.android.performance.internal.SpanProcessor;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import kotlin.jvm.functions.Function1;

public class TestSpanFactory {
    private long spanCount = 1L;

    final TestTimeoutExecutor timeoutExecutor = new TestTimeoutExecutor();

    public SpanImpl openSpan(SpanProcessor processor) {
        return newSpan(
                null,
                null,
                0L,
                null,
                null,
                spanCount++,
                0L,
                null,
                processor
        );
    }

    public SpanImpl namedSpan(String spanName) {
        return newSpan(
                spanName,
                null,
                System.currentTimeMillis(),
                null,
                null,
                0L,
                0L,
                null,
                NoopSpanProcessor.INSTANCE
        );
    }

    public SpanImpl newSpan(UUID traceId, long spanId) {
        return newSpan(
                null,
                null,
                System.currentTimeMillis(),
                null,
                traceId,
                spanId,
                0L,
                null,
                NoopSpanProcessor.INSTANCE
        );
    }

    public SpanImpl newSpan(
            String name,
            SpanKind kind,
            long startTime,
            Function1<Long, Long> endTime,
            UUID traceId,
            long spanId,
            long parentSpanId,
            String spanCategory,
            SpanProcessor processor
    ) {
        if (spanId <= 0L) spanId = spanCount++;
        if (name == null) name = "Test/Span" + spanId;
        if (kind == null) kind = SpanKind.INTERNAL;
        if (startTime <= 0L) startTime = spanCount;
        if (traceId == null) traceId = new UUID(0L, spanId);
        if (spanCategory == null) spanCategory = SpanCategory.CATEGORY_CUSTOM;

        SpanImpl span = new SpanImpl(
                name,
                spanCategory,
                kind,
                startTime,
                traceId,
                spanId,
                parentSpanId,
                true,
                timeoutExecutor,
                processor
        );

        if (endTime != null) {
            span.end(endTime.invoke(startTime));
        }

        return span;
    }

    public SpanImpl newSpan(SpanProcessor processor) {
        return newSpan(
                null,
                null,
                spanCount,
                t -> t + 10L,
                null,
                spanCount,
                0L,
                null,
                processor
        );
    }

    public List<SpanImpl> newSpans(int count, SpanProcessor processor) {
        List<SpanImpl> spans = new ArrayList<>();
        for (int i = 0; i < count; i++) {
            spans.add(newSpan(processor));
        }
        return spans;
    }
}

File: bugsnag-android-performance/src/test/java/com/bugsnag/android/performance/test/TestSpanFactory.kt
package com.bugsnag.android.performance.test

import com.bugsnag.android.performance.SpanKind
import com.bugsnag.android.performance.internal.SpanCategory
import com.bugsnag.android.performance.internal.SpanImpl
import com.bugsnag.android.performance.internal.SpanProcessor
import java.util.UUID

class TestSpanFactory {
    private var spanCount = 1L

    internal val timeoutExecutor = TestTimeoutExecutor()

    fun openSpan(processor: SpanProcessor): SpanImpl =
        newSpan(
            name = null,
            kind = null,
            startTime = 0L,
            endTime = null,
            traceId = null,
            spanId = spanCount++,
            parentSpanId = 0L,
            spanCategory = null,
            processor = processor,
        )

    fun namedSpan(spanName: String): SpanImpl =
        newSpan(
            name = spanName,
            kind = null,
            startTime = System.currentTimeMillis(),
            endTime = null,
            traceId = null,
            spanId = 0L,
            parentSpanId = 0L,
            spanCategory = null,
            processor = NoopSpanProcessor.INSTANCE,
        )

    fun newSpan(
        traceId: UUID,
        spanId: Long,
    ): SpanImpl =
        newSpan(
            name = null,
            kind = null,
            startTime = System.currentTimeMillis(),
            endTime = null,
            traceId = traceId,
            spanId = spanId,
            parentSpanId = 0L,
            spanCategory = null,
            processor = NoopSpanProcessor.INSTANCE,
        )

    @Suppress("LongParameterList")
    fun newSpan(
        name: String? = null,
        kind: SpanKind? = null,
        startTime: Long = 0L,
        endTime: ((Long) -> Long)? = null,
        traceId: UUID? = null,
        spanId: Long = 0L,
        parentSpanId: Long = 0L,
        spanCategory: SpanCategory? = null,
        processor: SpanProcessor = NoopSpanProcessor.INSTANCE,
    ): SpanImpl {
        val resolvedSpanId = if (spanId <= 0L) spanCount++ else spanId
        val resolvedName = name ?: "Test/Span$resolvedSpanId"
        val resolvedKind = kind ?: SpanKind.INTERNAL
        val resolvedStartTime = if (startTime <= 0L) spanCount else startTime
        val resolvedTraceId = traceId ?: UUID(0L, resolvedSpanId)
        val resolvedCategory = spanCategory ?: SpanCategory.CUSTOM

        val span =
            SpanImpl(
                resolvedName,
                resolvedCategory.category,
                resolvedKind,
                resolvedStartTime,
                resolvedTraceId,
                resolvedSpanId,
                parentSpanId,
                true,
                timeoutExecutor,
                processor,
            )

        if (endTime != null) {
            span.end(endTime.invoke(resolvedStartTime))
        }

        return span
    }

    fun newSpans(
        count: Int,
        processor: SpanProcessor,
    ): List<SpanImpl> {
        return List(count) { newSpan(processor = processor) }
    }
}
